### DISCUSSION TITLE
Web language support

### DISCUSSION THREAD

**Discussion by david-alm** [ralfbiedert/interoptopus#60] [Ideas]
Hey folks 👋 

I was wondering if there were any existing plans to build a web language interoptopus backend? WASM + Typescript would be an example of a relatively interoperable constellation.

Unfortunately, even the best existing approach to generating web bindings with wasm-bindgen involves a significant amount of pain and complexity when defining shims to existing crates.

Any thoughts or input are appreciated!

**Comment by ralfbiedert** [comment]
> I was wondering if there were any existing plans to build a web language interoptopus backend?

Hi, I don't have plans to do this myself any time soon, but PRs or work exploring this would be welcome. 

> Unfortunately, even the best existing approach to generating web bindings with wasm-bindgen involves a significant amount of pain and complexity when defining shims to existing crates.

Yes, but I think this will be similar here:

- Interoptopus is an "explicit FFI" crate where you need to design a C-level API.  Wrapping existing crates means (even for C# or Python) thinking about how such a "shim" / FFI layer would look like. 
- For WASM in particular, IIRC their type system is still relatively limited (w.r.t. numeric types, pointers, ...). So not only would you have to define an explicit FFI layer, but you'd also have to consider how to design that layer so it works well with limited primitives WASM offers. 

Not saying these are unsolvable, but some thought is required to make is working with such bindings nice.
